--- ./linux-5.4.155-bak/arch/x86/entry/syscalls/syscall_64.tbl	2021-10-22 08:55:16.131474221 +0800
+++ ./linux-5.4.155/arch/x86/entry/syscalls/syscall_64.tbl	2021-10-29 15:16:21.226020831 +0800
@@ -357,6 +357,7 @@
 433	common	fspick			__x64_sys_fspick
 434	common	pidfd_open		__x64_sys_pidfd_open
 435	common	clone3			__x64_sys_clone3/ptregs
+436 common  mysyscall       sys_mysyscall
 
 #
 # x32-specific system call numbers start at 512 to avoid cache impact
